src/sidebar/components/markdown-editor.js

```diff
@@ -134,69 +136,180 @@ ToolbarButton.propTypes = {
+ */
 function Toolbar({ isPreviewing, onCommand, onTogglePreview }) {
+  // Total number of elements in the toolbar minus 1.
+  const maxElementIndex = 9;
+
+  // Keep track of a roving index. The active roving tabIndex
+  // is set to 0, and all other elements are set to -1.
+  const [rovingElement, setRovingElement] = useState(0);
+
+  // An array of refs. Note we can't use `useRef` in a loop.
+  const elementsRefs = [];
+  for (let i = 0; i <= maxElementIndex; i++) {
+    elementsRefs.push(createRef());
+  }
+
+  /**
+   * Set the element to be focused and set it to be
+   * the active roving index.
+   *
+   * @param {number} index - Ordered index that matches the element
+   */
+  const setFocusedElement = index => {
+    setRovingElement(index);
+    try {
+      // preact component DOM elements are under 'base'
+      elementsRefs[index].current.base.focus();
+    } catch (e) {
+      // assume this is not a preact component, e.g. <a> tag
+      elementsRefs[index].current.focus();
+    }
+  };
+
+  /**
+   * Handles left and right arrow navigation as well as home and end
+   * keys so the user may navigate the toolbar without multiple tab stops.
+   */
+  const handleKeyDown = e => {
+    let lowerLimit = 0;
+    const upperLimit = maxElementIndex;
+    if (isPreviewing) {
+      // When isPreviewing is true, only allow navigation of
+      // the last 2 items.
+      lowerLimit = 8;
+    }
+    switch (e.key) {
+      case 'ArrowLeft':
+        if (rovingElement <= lowerLimit) {
+          setFocusedElement(upperLimit);
+        } else {
+          setFocusedElement(rovingElement - 1);
+        }
+        e.preventDefault();
+        break;
+      case 'ArrowRight':
+        if (rovingElement >= upperLimit) {
+          setFocusedElement(lowerLimit);
+        } else {
+          setFocusedElement(rovingElement + 1);
+        }
+        e.preventDefault();
+        break;
+      case 'Home':
+        setFocusedElement(lowerLimit);
+        e.preventDefault();
+        break;
+      case 'End':
+        setFocusedElement(upperLimit);
+        e.preventDefault();
+        break;
+    }
+  };
+
+  /**
+   * Returns the tab index value for a given element.
+   * Each element should be set to "-1" unless its the
+   * active roving index, in which case it will be "0"
+   *
+   * @param {number} elementIndex - Element's ordered index
+   * @return {string} `tabIndex` value
+   */
+  const getTabIndex = elementIndex => {
+    if (rovingElement === elementIndex) {
+      return '0';
+    } else {
+      return '-1';
+    }
+  };
+
   return (
     <div
       className="markdown-editor__toolbar"
       role="toolbar"
       aria-label="Markdown editor toolbar"
+      onKeyDown={handleKeyDown}
     >
       <ToolbarButton
         disabled={isPreviewing}
         iconName="format-bold"
         onClick={() => onCommand('bold')}
         shortcutKey={SHORTCUT_KEYS.bold}
+        ref={elementsRefs[0]}
```

Comment:
The simple dumb way of doing this is to put the link-age directly into the JSX like this as opposed to some sort of array of state that we reference. I think this way ends up being a bit more simple, but we  need to be careful about the index values because they are all hard coded here. e.g. if one changes, we need to change everyone after that.

Reply:
I did a bit more indirection if thats okay?, but same thing otherwise.